docs/2.10.0/docs/upgrade/smart-contract-upgrades.rst

```diff
@@ -2422,3 +2236,189 @@ recommend at least one test for your core workflows that follows this pattern:
+Queries for package-names will return all versions of a given contract, alongside the
+package-version and package-id for each contract.
+
+.. note::
+  If you still need to perform a query with an explicit package-id, you can either use
+  a previous version of PQS, or add the following filter predicate to your query:
+  ``SELECT \* FROM active('my_package:My.App:MyTemplate') WHERE package_id = 'my_package_id'``
+
+Given that PQS uses a document-oriented model for ledger content
+(JSONB), extensions to contract payloads are handled simply by returning
+the additional data in the blob.
+
+Daml Shell
+----------
+
+Daml Shell builds on PQS by providing a shell interface to inspect the
+ledger using package-name to view all versions of contracts, in an
+integrated way.
+
+Daml-Script
+-----------
+
+Daml 2.10 introduces a new version of Daml Script, which can be used by
+depending on ``daml-script-lts`` in your ``daml.yaml``, as you will have seen
+in `Writing your first upgrade <#writing-your-first-upgrade>`__. This version of Daml Script
+supports upgrades over the Ledger API.
+
+All commands and queries in this version of Daml Script now use
+upgrades/downgrades automatically, to ensure that the correct versions
+of choices are exercises, and correct payloads are returned.
+
+The following additional functionality is available for more advanced
+uses of SCU.
+
+**Exact commands**
+
+Each of the four submission commands now has an "exact" variant, of the
+forms ``createExactCmd``, ``exerciseExactCmd``, ``exerciseByKeyExactCmd`` and
+``createAndExerciseExactCmd``. These commands force the participant to
+use the exact version of the package that your script uses, this is most
+useful when you want to be absolutely certain of the choice code you are
+calling. Note that exact and non-exact commands can be mixed in the same
+submission.
+
+.. _daml-script-package-preference:
+
+**Package Preference**
+
+A submission can specify a `package preference <#dynamic-package-resolution-in-ledger-api-queries>`__,
+as a list of package IDs:
+
+.. code:: daml
+
+  (actAs alice <> packagePreference [myPackageId]) `submitWithOptions` createCmd ...
+
+Note the use of ``submitWithOptions : SubmitOptions -> Commands a -> Script a``.
+You can build ``SubmitOptions`` by combining the ``actAs`` and ``packagePreference`` functions with ``<>``.
+
+The full list of builders for ``SubmitOptions`` is as follows:
+
+.. code:: daml
+
+  -- `p` can be `Party`, `[Party]`, etc.
+  actAs : IsParties p => p -> SubmiOptions
+  readAs : IsParties p => p -> SubmitOptions
+
+  disclose : Disclosure -> SubmitOptions
+  discloseMany : [Disclosure] -> SubmitOptions
+
+  newtype PackageId = PackageId Text
+  packagePreference : [PackageId] -> SubmitOptions
+
+A ``PackageId`` can be hard-coded in your script, which must be updated whenever the package changes. Otherwise,
+it can be provided using the ``--input-file`` flag of the ``daml script`` command line tool.
```

Comment:
```suggestion
A ``PackageId`` can be hard-coded in your script, in which case it must be updated whenever the package changes. Otherwise,
it can be provided using the ``--input-file`` flag of the ``daml script`` command line tool.
```